### PR TITLE
Add a script to verify safetensor checkpoint

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Verify the converted safetensor checkpoint (GCS or local) matches the remote HuggingFace checkpoint reference. 
* Add the standalone script
* Move the helper functions to the `utils.py`

We may enable lazy mode as follow up to save memory.

# Tests

* Successfully run and load from gcs: [test](https://paste.googleplex.com/5346139610152960)
* Successfully run and load from local: [test](https://paste.googleplex.com/5757938390138880)
* Mismatch report: key mismatch [test](https://paste.googleplex.com/5849828544348160)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/utils/ckpt_conversion/verify_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+def get_hf_model_state_dict(model_id: str, token: str) -> Dict[str, np.ndarray]:
+  """Loads the HuggingFace model state dict and converts to numpy."""
+  max_logging.log(f"Loading reference model from HuggingFace: {model_id}...")
+
+  hf_model = get_hf_model(model_id, token)
+  state_dict = hf_model.state_dict()
+  numpy_state_dict = {k: v.numpy() for k, v in state_dict.items()}
+
+  del hf_model
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/verify_checkpoint.py]: nit: will scope cleanup automatically?

File: src/MaxText/utils/ckpt_conversion/verify_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+"""
+Verify the converted safetensor checkpoint (GCS or local) matches the remote HuggingFace checkpoint reference. 
+
+Usage for scanned checkpoint:
+JAX_PLATFORMS=cpu python3 -m MaxText.utils.ckpt_conversion.verify_checkpoint src/MaxText/configs/base.yml \
+    model_name=<maxtext_model_name> \
+    use_multimodal=false \
+    hf_access_token=<your_hf_token> \
+    scan_layers=true hardware=cpu \
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/verify_checkpoint.py]: where is scan_layers used?

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
+
+  # Parse script-specific arguments
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      "--checkpoint_path",
+      type=str,
+      default="",
+      required=True,
+      help="The path to the converted safetensors checkpoint (e.g. gs://bucket/path or /local/path)",
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py:228]: Thanks for addressing both cases!

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  loaded_tensors = load_safetensors(file_bytes)
+
+  # Convert to Numpy
+  shard_dict = {}
+  for key, tensor in loaded_tensors.items():
+    shard_dict[key] = tensor.numpy()
+
+  del file_bytes
+  del loaded_tensors
+  return shard_dict
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: nit: will scope cleanup automatically?

File: src/MaxText/utils/ckpt_conversion/verify_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Verify the converted safetensor checkpoint (GCS or local) matches the remote HuggingFace checkpoint reference. 
+
+Usage for scanned checkpoint:
+JAX_PLATFORMS=cpu python3 -m MaxText.utils.ckpt_conversion.verify_checkpoint src/MaxText/configs/base.yml \
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/verify_checkpoint.py]: maybe use a more specific name? verify_checkpoint is somewhat generic (e.g., compare_hf_checkpoint)

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  model_name = config.model_name
+  if model_name not in HF_IDS:
+    raise ValueError(f"Unsupported model name: {model_name}. " f"Supported: {list(HF_IDS.keys())}")
+
+  model_id = HF_IDS[model_name]
+  hf_token = config.hf_access_token
+
+  # 1. Load Reference (HuggingFace)
+  t0 = time.perf_counter()
+  hf_state_dict = get_hf_model_state_dict(model_id, hf_token)
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: nice to have but not required: option to load from local reference

**Review by shuningjin — COMMENTED** [review]
Thank you for adding this utility! Looks good overall.

File: src/MaxText/utils/ckpt_conversion/verify_checkpoint.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/verify_checkpoint.py]: Good catch! We actually never need it!

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py (hunk 6 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Could you explain a little bit more? I think the goal is directly compare against the source of truth in HF hub/remote.

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+  t1 = time.perf_counter()
+  max_logging.log(f"⏱️ HuggingFace model loaded in {(t1 - t0) / 60:.2f} minutes")
+
+  # 2. Load Candidate (GCS or Local)
+  t0 = time.perf_counter()
+  cand_state_dict = load_safetensors_generic(test_args.checkpoint_path, test_args.max_workers)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Good call, thanks! Updated.

File: src/MaxText/utils/ckpt_conversion/verify_checkpoint.py (hunk 5 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/verify_checkpoint.py]: Cool! Updated to `compare_hf_ckpt` for shorter name.

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py (hunk 4 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Good call. Removed them as we are at the end of a function scope

File: src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py (hunk 6 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Discussed offline. We have low-precision checkpoints that may not directly comparable in fp8, fp4, etc. We will have a follow up PR to enable reading from gcs/local file options for reference checkpoint.

Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Actually, updated the feature in this PR. Also tested:

1) candidate ckpt (local) to remote HF ckpt: https://paste.googleplex.com/5400855413129216
2) candidate ckpt (local) to reference ckpt (local): https://paste.googleplex.com/4699355451883520

Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/compare_hf_ckpt.py]: Thanks for adding this!

**Review by hengtaoguo — APPROVED** [review]
Thanks! I also wonder if this is equivalent to the forward logits check, which also requires all layers identical to have matching results. It's great to have a double insurance though.

**Comment by RissyRan** [comment]
> Thanks! I also wonder if this is equivalent to the forward logits check, which also requires all layers identical to have matching results. It's great to have a double insurance though.

Thanks Hengtao! No, it's not the same. Forward logit test is testing orbax checkpoint loading into maxtext against with HF. However this test is used specifically for safetensor checkpoint (converted from to_huggingface) against with reference HF checkpoint.
